### PR TITLE
plan, executor: make Analyze a Plan

### PR DESCRIPTION
**Pull request by lamxTyler**
This PR mainly do two things:
* make Analyze a Plan, not PhysicalPlan
* remove XSelectIndexExec's dependency on PhyscialIndexScan

PTAL @shenli @coocood @hanfei1991 @winoros

### CODE REVIEW

File: executor/builder.go [hunk 1]
```diff
@@ -743,22 +751,78 @@ func (b *executorBuilder) buildCache(v *plan.Cache) Executor {
 	}
 }
 
+func (b *executorBuilder) buildSimpleTableScan(tblInfo *model.TableInfo, cols []*model.ColumnInfo) Executor {
```
Comment (hanfei1991) [inline executor/builder.go]: buildTableScanForAnalyze

File: executor/builder.go [hunk 2]
```diff
@@ -743,22 +751,78 @@ func (b *executorBuilder) buildCache(v *plan.Cache) Executor {
...
+		supportDesc: supportDesc,
+		table:       table,
+		schema:      schema,
+		Columns:     cols,
+		ranges:      ranges,
+	}
+	return e
+}
+
+func (b *executorBuilder) buildSimpleIndexScan(tblInfo *model.TableInfo, idxInfo *model.IndexInfo) Executor {
```
Comment (hanfei1991) [inline executor/builder.go]: ditto

File: executor/builder.go [hunk 3]
```diff
@@ -743,22 +751,78 @@ func (b *executorBuilder) buildCache(v *plan.Cache) Executor {
...
+	table, _ := b.is.TableByID(tblInfo.ID)
+	client := b.ctx.GetClient()
+	supportDesc := client.SupportRequestType(kv.ReqTypeSelect, kv.ReqSubTypeDesc)
+	schema := expression.NewSchema(expression.ColumnInfos2Columns(tblInfo.Name, cols)...)
+	ranges := []types.IntColumnRange{{math.MinInt64, math.MaxInt64}}
+	e := &XSelectTableExec{
+		tableInfo:   tblInfo,
+		ctx:         b.ctx,
+		startTS:     startTS,
+		supportDesc: supportDesc,
```
Comment (hanfei1991) [inline executor/builder.go]: Do not need to set it

File: executor/builder.go [hunk 4]
```diff
@@ -743,22 +751,78 @@ func (b *executorBuilder) buildCache(v *plan.Cache) Executor {
...
+	for i, col := range idxInfo.Columns {
+		cols[i] = tblInfo.Columns[col.Offset]
+	}
+	schema := expression.NewSchema(expression.ColumnInfos2Columns(tblInfo.Name, cols)...)
+	idxRange := &types.IndexRange{LowVal: []types.Datum{types.MinNotNullDatum()}, HighVal: []types.Datum{types.MaxValueDatum()}}
+	scanConcurrency := b.ctx.GetSessionVars().IndexSerialScanConcurrency
+	e := &XSelectIndexExec{
+		tableInfo:       tblInfo,
+		ctx:             b.ctx,
+		supportDesc:     supportDesc,
```
Comment (hanfei1991) [inline executor/builder.go]: Ditto

**Comment by hanfei1991** [comment]
LGTM

File: plan/physical_plans.go [hunk 5]
```diff
@@ -56,7 +56,6 @@ var (
 	_ PhysicalPlan = &Limit{}
 	_ PhysicalPlan = &Show{}
 	_ PhysicalPlan = &Insert{}
-	_ PhysicalPlan = &Analyze{}
```
Comment (shenli) [inline plan/physical_plans.go]: So which interface does Analyze implemented?

Comment (hanfei1991) [inline plan/physical_plans.go]: plan.Plan

**Comment by shenli** [comment]
Why should make Analyze  a Plan instead of a PhysicalPlan?

**Comment by hanfei1991** [comment]
@shenli unnecessary

**Comment by shenli** [comment]
LGTM

**Pull request merged**